jxl/src/api/gain_map.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//! Support for ISO 21496-1 gain maps in JPEG XL.
+//!
+//! Gain maps allow encoding HDR images with an SDR fallback, where the gain map
+//! describes how to convert from SDR to HDR representation.
+
+use crate::error::{Error, Result};
+use crate::headers::color_encoding::ColorEncoding;
+
+/// A gain map bundle as defined by ISO 21496-1.
+///
+/// This structure contains all data from a `jhgm` box in a JPEG XL container,
+/// including the gain map metadata, optional color encoding, ICC profile, and
+/// the gain map codestream itself.
+#[derive(Debug, Clone)]
+pub struct GainMapBundle {
+    /// Version of the gain map format (currently always 0).
+    pub jhgm_version: u8,
+
+    /// ISO 21496-1 metadata blob (binary format).
+    ///
+    /// This contains parameters like gain_map_min, gain_map_max, gamma, offset, etc.
+    /// The format is defined by the ISO 21496-1 standard.
+    pub gain_map_metadata: Vec<u8>,
+
+    /// Optional color encoding for the gain map.
+    ///
+    /// If present, this describes the color space of the gain map image.
+    /// If None, the gain map uses the same color encoding as the base image.
+    pub color_encoding: Option<ColorEncoding>,
+
+    /// Alternative ICC profile for the gain map (may be Brotli-compressed).
+    ///
+    /// This is used when the color encoding cannot be fully described by the
+    /// JPEG XL ColorEncoding structure.
+    pub alt_icc: Vec<u8>,
+
+    /// The gain map image data as a JPEG XL naked codestream.
```

Comment:
Doesn't have to be naked. It can be an arbitrary embedded JPEG XL file, also one using the container (but it is not allowed to itself contain a gain map box). Originally it was required to be a naked codestream, but allowing embedded containers makes it possible to have a gain map with `jbrd`, in case the gain map is itself a recompressed JPEG and you want to be able to reconstruct that JPEG bitstream exactly.